source/common/router/header_parser.cc

```diff
@@ -287,5 +294,26 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
   }
 }
 
+void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
+                                   const StreamInfo::StreamInfo* stream_info) const {
```

Comment:
Could we use a null StreamInfo rather than duplicate the add/remove logic?

Reply:
Sorry, do you mean to have something like `NullStreamInfoImpl`?

Reply:
The only problem with that probably if we have an input that creates `StreamInfoHeaderFormatter` (or `CompoundHeaderFormatter`) we need to always act as `PlainHeaderFormatter`.

For example: given:

- name: key
   value: %DOWNSTREAM_LOCAL_ADDRESS%

This creates `StreamInfoHeaderFormatter`, if we have `NullStreamInfoImpl` we need to provide a default value for `stream_info.downstreamLocalAddress()`, which probably could be misleading?

Reply:
OK, fair enough, but could we merge the control flow so that we have only one internal evaluateHeadersInt and this can branch at the leaves on what to do?

Reply:
Have done a bit of refactoring, hope this aligns with your expectation :).